### PR TITLE
Fix the way to read the blob for GetSchema feature

### PR DESCRIPTION
**Pull request by kjcho-msft**

### CODE REVIEW

File: Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs [hunk 1]
```diff
@@ -152,33 +153,118 @@ public static async Task<List<string>> GetLastModifiedBlobContentsInBlobPath(str
...
+                    throw new GeneralException("Token in the blob path should be a data time format. e.g. {yyyy-MM-dd}");
+                }
+
+                path = path.Replace(m.Value, @"(\w+)", StringComparison.InvariantCulture);
+            }
+
+            return path;
+        }
+
+        private static string NormalizeBlobPath(string path)
```
Comment (carlbrochu) [inline Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs]: Function headers comments

File: Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs [hunk 2]
```diff
@@ -152,33 +153,118 @@ public static async Task<List<string>> GetLastModifiedBlobContentsInBlobPath(str
...
+        {
+            if (!Uri.TryCreate(blobUri, UriKind.Absolute, out var uri))
+            {
+                return blobUri;
+            }
+
+            var prefix = "";
+            foreach (var seg in uri.Segments)
+            {
+                if (seg.StartsWith("%7B")) // "{", '\u007B'
```
Comment (carlbrochu) [inline Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs]: comment?

File: Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs (hunk 1 above)
Comment (kjcho-msft) [reply Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs]: add header comments

File: Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs (hunk 2 above)
Comment (kjcho-msft) [reply Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs]: added comments

File: Services/DataX.Flow/DataX.Flow.SchemaInference/Blob/BlobMessageBus.cs [hunk 3]
```diff
@@ -34,6 +34,7 @@ public BlobMessageBus(List<FlowGuiInputBatchInput> batchInputs, ILogger logger)
         public async Task<EventsData> GetSampleEvents(int seconds)
         {
             EventsData eventsData = new EventsData();
+            const int numberOfDocumentsToRead = 500;
```
Comment (carlbrochu) [inline Services/DataX.Flow/DataX.Flow.SchemaInference/Blob/BlobMessageBus.cs]: does this affect streaming as well?  can this be a parameter to the function

File: Services/DataX.Flow/DataX.Flow.SchemaInference.Tests/SchemaUtilityTests.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    [TestClass]
+    public class SchemaUtilityTests
+    {
+        [TestMethod]
+        public void BlobPathPrefixTest()
+        {
+            List<Tuple<string, string, string, string>> blobPaths = new List<Tuple<string, string, string, string>>
+            {
+                new Tuple<string, string, string, string>(
+                    @"wasbs://mycontainer@mysa.blob.core.windows.net/mypath/mypath2/mypath3/{yyyy}/{MM}/{dd}",
```
Comment (vijayupadya) [inline Services/DataX.Flow/DataX.Flow.SchemaInference.Tests/SchemaUtilityTests.cs]: Does this pattern still work with the regex updates?

File: Services/DataX.Flow/DataX.Flow.SchemaInference/Blob/BlobMessageBus.cs (hunk 3 above)
Comment (kjcho-msft) [reply Services/DataX.Flow/DataX.Flow.SchemaInference/Blob/BlobMessageBus.cs]: No this code path is only for blob input.

File: Services/DataX.Flow/DataX.Flow.SchemaInference.Tests/SchemaUtilityTests.cs (hunk 4 above)
Comment (kjcho-msft) [reply Services/DataX.Flow/DataX.Flow.SchemaInference.Tests/SchemaUtilityTests.cs]: No regex update was needed. The current logic just works with this {yyyy/MM/dd}. Add a couple tests to test {yyyy/MM/dd} and {yyyy/MM/dd}/test.

File: Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs [hunk 5]
```diff
@@ -141,51 +163,178 @@ private static CloudBlockBlob GetBlobReference(string connectionString, string b
...
+                    using (var stream = blob.OpenReadAsync().Result)
+                    {
+                        using (var sr = new StreamReader(stream))
+                        {
+                            int offset = 0;
+                            while (blob.Properties.Length > offset)
+                            {
+                                // We can't read the entire blob if a blob is too big (1 > GB)
+                                // So we just read it line by line until we have enough information (500 rows) to generate schema
+                                var line = sr.ReadLineAsync().Result;
```
[the remark this answers is not in the document]
Comment (kjcho-msft) [reply Services/DataX.Utilities/DataX.Utilities.Blob/BlobHelper.cs]: Good point. made it an async call.

**Pull request merged**